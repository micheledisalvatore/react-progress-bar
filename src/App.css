.app {
  text-align: center;
}

.app__change-step {
  background-color: #5131D7;
  border-radius: 5px;
  border: 1px solid #5131D7;
  color: white;
  font-size: 1rem;
  font-weight: bold;
  height: 2rem;
  margin: 0 10px;
  width: 2rem;
}

.app__change-step:disabled {
  background-color: #d7ceff;
  border-color: #d7ceff;
}
