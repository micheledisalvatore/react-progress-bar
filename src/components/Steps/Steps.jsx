import React from 'react';
import PropTypes from 'prop-types';

import Step from '../Step';

import './Steps.css';

export const Steps = ({total, current, onClick}) => (
  <div className="steps">
    {[...(new Array(total))].map((e, i) => (
      <Step
        key={i}
        title={`Step ${i + 1}`}
        step={i + 1}
        isFirst={i === 0}
        isActive={i < current}
        onClick={onClick}
      />
    ))}
  </div>
);

Steps.propTypes = {
  total: PropTypes.oneOf([2,3,4,5]).isRequired,
  current: PropTypes.oneOf([1,2,3,4,5]).isRequired,
  onClick: PropTypes.func.isRequired,
};