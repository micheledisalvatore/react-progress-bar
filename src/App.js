import React, {Component} from 'react';

import Steps from './components/Steps';

import './App.css';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      steps: 2,
      currentStep: 1,
    }
  }

  onStepClick = nextStep => {
    this.setState(({currentStep}) => {
      if (nextStep === currentStep + 1 || nextStep === currentStep - 1) {
        return {currentStep: nextStep};
      }
    })
  };

  get canIncrease() {
    return this.state.steps < 5
  }

  get canDecrease() {
    return this.state.steps > 2
  }

  changeTotalSteps = isIncreasing => () => {
    this.setState(({steps, currentStep}) => {
      let newSteps = steps;
      if (this.canIncrease && isIncreasing) {
        newSteps += 1;
      } else if (this.canDecrease && !isIncreasing) {
        newSteps -= 1;
      }
      let newCurrentStep = Math.min(currentStep, newSteps);

      return {
        steps: newSteps,
        currentStep: newCurrentStep,
      };
    })
  };

  render() {
    const {steps, currentStep} = this.state;
    return (
      <div className="app">
        <div>
          <h1>Change here the steps number</h1>
          <button name="decrease" onClick={this.changeTotalSteps(false)} className="app__change-step" disabled={!this.canDecrease}>-</button>
          <button name="increase" onClick={this.changeTotalSteps(true)} className="app__change-step" disabled={!this.canIncrease}>+</button>
        </div>
        <Steps total={steps} current={currentStep} onClick={this.onStepClick}/>
        <div className="app__credits">
          To check the source code, please visit
          the <a href="https://gitlab.com/michele.disalvatore/react-progress-bar" target="_blank" rel="noopener noreferrer">GitLab repository</a>
        </div>
      </div>
    );
  }
}

export default App;
