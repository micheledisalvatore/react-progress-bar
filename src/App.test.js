import React from 'react';
import { shallow } from 'enzyme';

import Steps from './components/Steps';

import App from './App';

describe('Given an App component', () => {
  describe('when it is rendered', () => {
    const component = <App />;
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(component);
    });

    it('should match the snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('should render a Steps component', () => {
      expect(wrapper.find(Steps)).toBePresent();
    });

    it('should render Steps with total prop set to 2', () => {
      expect(wrapper.find(Steps)).toHaveProp('total', 2);
    });

    it('should render Steps with current prop set to 1', () => {
      expect(wrapper.find(Steps)).toHaveProp('current', 1);
    });

    it('should render 2 buttons component', () => {
      expect(wrapper.find('button').length).toBe(2);
    });

    describe('and a user adds 1 more steps', () => {
      beforeEach(() => {
        wrapper.find('[name="increase"]').simulate('click');
      });

      it('should render Steps with total prop set to 3', () => {
        expect(wrapper.find(Steps)).toHaveProp('total', 3);
      });

      it('should render Steps with current prop set to 1', () => {
        expect(wrapper.find(Steps)).toHaveProp('current', 1);
      });

      describe('and the user adds 2 more steps', () => {
        beforeEach(() => {
          wrapper.find('[name="increase"]').simulate('click');
          wrapper.find('[name="increase"]').simulate('click');
        });

        it('should render Steps with total prop set to 5', () => {
          expect(wrapper.find(Steps)).toHaveProp('total', 5);
        });

        describe('and the user clicks on a next valid step', () => {
          beforeEach(() => {
            wrapper.find(Steps).simulate('click', 2);
          });

          it('should render Steps with current prop set to 2', () => {
            expect(wrapper.find(Steps)).toHaveProp('current', 2);
          });

          describe('and the user clicks on a next valid step', () => {
            beforeEach(() => {
              wrapper.find(Steps).simulate('click', 3);
            });

            it('should render Steps with current prop set to 3', () => {
              expect(wrapper.find(Steps)).toHaveProp('current', 3);
            });

            describe('and the user clicks to remove 3 steps', () => {
              beforeEach(() => {
                wrapper.find('[name="decrease"]').simulate('click');
                wrapper.find('[name="decrease"]').simulate('click');
                wrapper.find('[name="decrease"]').simulate('click');
              });

              it('should render Steps with total prop set to 2', () => {
                expect(wrapper.find(Steps)).toHaveProp('total', 2);
              });

              it('should render Steps with current prop set to 2', () => {
                expect(wrapper.find(Steps)).toHaveProp('current', 2);
              });
            })
          });
        });

        describe('and the user clicks on a next NOT valid step', () => {
          beforeEach(() => {
            wrapper.find(Steps).simulate('click', 3);
          });

          it('should render Steps with current prop NOT set to 3', () => {
            expect(wrapper.find(Steps)).not.toHaveProp('current', 3);
          });
        });

        describe('and the user adds 1 more step', () => {
          beforeEach(() => {
            wrapper.find('[name="increase"]').simulate('click');
          });

          it('should render Steps with total prop NOT set to 6', () => {
            expect(wrapper.find(Steps)).not.toHaveProp('total', 6);
          });
        });
      });
    })
  });
});