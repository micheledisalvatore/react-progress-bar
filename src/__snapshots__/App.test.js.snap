// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an App component when it is rendered should match the snapshot 1`] = `
<div
  className="App"
>
  <div>
    <h1>
      Change here the steps number
    </h1>
    <button
      name="decrease"
      onClick={[Function]}
    >
      -
    </button>
    <button
      name="increase"
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <Unknown
    current={1}
    onClick={[Function]}
    total={2}
  />
</div>
`;
